File: algo_3_two_points/L345.cpp
//
// Created by LiHeng on 2020/6/3.
//

// 345 反转字符串中的元音字母
#include <iostream>
#include <unordered_set>

using namespace std;
class Solution {
public:
    string reverseVowels(string s) {
        if(s == "") return "";
        unordered_set<char> st = {'a','e','i','o','u','A', 'E', 'I', 'O', 'U'};
        int i=0, j=s.size()-1;
        while(i<j){
            if(st.find(s[i]) == st.end())//左边字母不是元音
                ++i;
            else if(st.find(s[j]) == st.end())//右边字母不是元音
                --j;
            else
                swap(s[i++],s[j--]);
        }
        return s;
    }
};

File: python-algorithm/algo_03_two_points/L345.py
#--coding:utf-8--

class Solution:
    def reverseVowels(self, s: str) -> str:
        s_l = [c for c in s] # 字符串不可变，转成列表表达
        tt = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']
        lt, rt = 0, len(s)-1
        while lt < rt:
            while lt < rt and s_l[lt] not in tt:
                lt += 1
            while lt < rt and s_l[rt] not in tt:
                rt -= 1
            s_l[lt], s_l[rt] = s_l[rt], s_l[lt]
            lt += 1
            rt -= 1
        return ''.join(s_l)